lightning/src/ln/channelmanager.rs

```diff
@@ -3935,7 +3939,7 @@ where
 			log_info!(self.logger, "Attempted to claim an incomplete payment which no longer had any available HTLCs!");
 			return;
 		}
-		if claimable_amt_msat != expected_amt_msat.unwrap() {
+		if claimable_amt_msat < expected_amt_msat.unwrap() {
```

Comment:
Hm, it does look redundant. I tried taking it out to see and it seems it's relevant in the functional test `test_double_partial_claim`, in the case where HTLCs timeout before the user claims their funds, then the sender resends only part of the payment, and *then* the user tries to process the original `PaymentClaimable` event and claim the payment. I guess were you talking about if the check was necessary or something else?

Reply:
Oh gotcha thanks for the clarification

Reply:
Agh my bad, sorry for misunderstanding again, what I thought you were talking about in this comment was that `claimable_amt_msat` had been changed to be the sum of the onion values (as `htlc.value` had been changed to be `outgoing_amt_msat`, i.e. `amt_to_forward` from the onion), so I changed `htlc.value` back to be the received amount (and added `sender_intended_value` for the onion amount to use in MPP set calculation).

IIUC I don't think `total_msat` was changed in this PR, just the validation against it, so I believe `htlc.total_msat`/`expected_amt_msat` was always the value in the onion?

Now that I get what you're saying however, I agree this check would make more sense if `expected_amt_msat` matched the amount reported in `PaymentClaimable`. Happy to add this here, or in a followup.

Reply:
> `total_msat` is still calculated as `htlc.total_msat` for MPP and `outgoing_amt_msat` for non-MPP, the problem is `outgoing_amt_msat` has changed here - its now the onion data rather the incoming amount. But, more importantly, the event has changed - the event now tells the user about the actual amount,

Okay, *noww* I see. Thank you so much for explaining!

> Let's just do it here?

On it!

Reply:
Thinking about this again, should I actually just change the amount reported in `PaymentClaimable` back to be the amount from the onion (i.e. `htlc.total_msat` for MPP and `outgoing_amt_msat` for non-MPP)? I'm thinking that will accomplish we want here: when claiming funds, `expected_amt_msat` will represent what was reported to the user in `PaymentClaimable`. The downside with this is that the event reports the amount from the invoice, while the actual amount may overshoot that.

I'm asking this because as I was trying to change `expected_amt_msat` to match `PaymentClaimable` now that `PaymentClaimable` reports the actual amount received, I'm realizing it introduces this case: a node overpays the original amount, `PaymentClaimable` is generated with the actual received amount, one/some of the HTLCs timeout before the receiver handles the event, now the sender resends HTLCs however now these HTLCs pay enough for the invoice but below `expected_amt_msat`, i.e. the amount previously reported in `PaymentClaimable`, so it refuses to claim. I'm thinking it makes more sense to be able to claim as long as we've received at least the amount originally requested in the invoice, and `PaymentClaimable` should just report that and sometimes the user will end up claiming a little extra?

Reply:
> We could also report the sender-reported amount to the user if we want, but I think that's confusing (its neither the amount they received nor the amount they specified in the invoice...but something else).

Ah, true.

> That's a good point, but I don't think we need to worry too much about this. If it causes payment failure, so be it - its generally not something that should ever happen HTLC timeouts after receiving them should be relatively rare, only when they were super tight to begin with and we get a block quickly thereafter. We could also consider making it arbitrarily rare by adding an extra block or two of headroom when accepting the HTLC vs when we go to time it out (I think I'll probably do that in a followup as a part of https://github.com/lightningdevkit/rust-lightning/issues/2017 anyway).

Okay gotcha, sounds good :)